server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/managable/cassandra/FirstEnqueuedMailDao.java

```diff
@@ -0,0 +1,120 @@
+ ****************************************************************/
+
+package org.apache.james.queue.rabbitmq.managable.cassandra;
+
+import static com.datastax.driver.core.querybuilder.QueryBuilder.bindMarker;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.eq;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.set;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.update;
+import static org.apache.james.queue.rabbitmq.managable.cassandra.FirstEnqueuedMailTable.QUEUE_NAME;
+import static org.apache.james.queue.rabbitmq.managable.cassandra.FirstEnqueuedMailTable.TABLE_NAME;
+import static org.apache.james.queue.rabbitmq.managable.cassandra.FirstEnqueuedMailTable.TIME_RANGE_START;
+
+import java.util.Date;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.queue.rabbitmq.MailQueueName;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.google.common.annotations.VisibleForTesting;
+
+public class FirstEnqueuedMailDao {
+
+    private final CassandraAsyncExecutor executor;
+    private final PreparedStatement selectOne;
+    private final PreparedStatement updateOne;
+    private final CassandraManageableMailQueueConfiguration mailQueueConfiguration;
+    @VisibleForTesting final Map<MailQueueName, Integer> queueUpdatePace;
+
+    @Inject
+    public FirstEnqueuedMailDao(Session session, CassandraManageableMailQueueConfiguration mailQueueConfiguration) {
+        this.executor = new CassandraAsyncExecutor(session);
+
+        this.selectOne = prepareSelectOne(session);
+        this.updateOne = prepareUpdate(session);
+        this.mailQueueConfiguration = mailQueueConfiguration;
+        this.queueUpdatePace = new ConcurrentHashMap<>();
+    }
+
+    private PreparedStatement prepareSelectOne(Session session) {
+        return session.prepare(select()
+                .from(TABLE_NAME)
+                .where(eq(QUEUE_NAME, bindMarker(QUEUE_NAME))));
+    }
+
+    private PreparedStatement prepareUpdate(Session session) {
+        return session.prepare(update(TABLE_NAME)
+                .with(set(TIME_RANGE_START, bindMarker(TIME_RANGE_START)))
+                .where(eq(QUEUE_NAME, bindMarker(QUEUE_NAME))));
+    }
+
+    public CompletableFuture<Void> updateEnqueuedMail(MailQueueName mailQueueName) {
+        updateCounter(mailQueueName);
+
+        if (isAbleToUpdateEnqueuedTable(mailQueueName)) {
+            return executor.executeVoid(updateOne.bind()
+                .setTimestamp(TIME_RANGE_START, new Date())
```

Comment:
No! We need to browse from previous known point until we find an email that is still in the queue...